test/shared/src/main/scala/zio/test/mock/MockRandom.scala

```diff
@@ -26,146 +30,141 @@ trait MockRandom extends Random {
+    import Mock._
 
-    def feedDoubles(doubles: Double*): UIO[Unit] =
-      randomState.update(data => data.copy(doubles = doubles.toList ::: data.doubles)).unit
+    val nextBoolean: UIO[Boolean] =
+      next(1).map(_ != 0)
 
-    def feedFloats(floats: Float*): UIO[Unit] =
-      randomState.update(data => data.copy(floats = floats.toList ::: data.floats)).unit
-
-    def feedLongs(longs: Long*): UIO[Unit] =
-      randomState.update(data => data.copy(longs = longs.toList ::: data.longs)).unit
-
-    def feedChars(chars: Char*): UIO[Unit] =
-      randomState.update(data => data.copy(chars = chars.toList ::: data.chars)).unit
-
-    def feedStrings(strings: String*): UIO[Unit] =
-      randomState.update(data => data.copy(strings = strings.toList ::: data.strings)).unit
-
-    def feedBytes(bytes: Chunk[Byte]*): UIO[Unit] =
-      randomState.update(data => data.copy(bytes = bytes.toList ::: data.bytes)).unit
-
-    val clearInts: UIO[Unit] =
-      randomState.update(data => data.copy(integers = List.empty)).unit
-
-    val clearBooleans: UIO[Unit] =
-      randomState.update(data => data.copy(booleans = List.empty)).unit
-
-    val clearDoubles: UIO[Unit] =
-      randomState.update(data => data.copy(doubles = List.empty)).unit
-
-    val clearFloats: UIO[Unit] =
-      randomState.update(data => data.copy(floats = List.empty)).unit
-
-    val clearLongs: UIO[Unit] =
-      randomState.update(data => data.copy(longs = List.empty)).unit
-
-    val clearChars: UIO[Unit] =
-      randomState.update(data => data.copy(chars = List.empty)).unit
-
-    val clearStrings: UIO[Unit] =
-      randomState.update(data => data.copy(strings = List.empty)).unit
-
-    val clearBytes: UIO[Unit] =
-      randomState.update(data => data.copy(bytes = List.empty)).unit
-
-    private def nextRandom[T](shift: Data => (T, Data)) =
+    val nextDouble: UIO[Double] =
       for {
-        data            <- randomState.get
-        (next, shifted) = shift(data)
-        _               <- randomState.update(_ => shifted)
-      } yield next
-
-    private def shiftBooleans(data: Data) =
-      (
-        data.booleans.headOption.fold(MockRandom.defaultBoolean)(identity),
-        data.copy(booleans = shiftLeft(data.booleans))
-      )
-
-    private def shiftIntegers(data: Data) =
-      (
-        data.integers.headOption.fold(MockRandom.defaultInteger)(identity),
-        data.copy(integers = shiftLeft(data.integers))
-      )
-
-    private def shiftIntWithLimit(limit: Int)(data: Data) = {
-      val next = data.integers.headOption.fold(MockRandom.defaultInteger)(identity)
-      (Math.min(limit, next), data.copy(integers = shiftLeft(data.integers)))
+        i1 <- next(26)
+        i2 <- next(27)
+      } yield ((i1.toLong << 27) + i2) / (1L << 53).toDouble
+
+    def nextBytes(length: Int): UIO[Chunk[Byte]] = {
+      //  Our RNG generates 32 bit integers so to maximize efficieny we want to
```

Comment:
Is this code original or derived? If derived, we should include credit, copyright notice, and make sure the license is Apache 2 compliant.

Reply:
It is based on `java.util.Random`. It looks like that is available under the GNU General Public License version 2. Does that work? Is there an example of including that information somewhere else in the project I could work off of?

Reply:
Unfortunately, we cannot include any GPL code. BSD/MIT/Apache 2 are all good licenses and we can incorporate that code with proper attribution and copyright notices.

ZIO is mostly original but there is some derived stuff here:

<https://github.com/zio/zio/blob/master/core/shared/src/main/scala/zio/Semaphore.scala>

Reply:
Okay, I could use some guidance on how to proceed here. Right now I'm trying to generate the same sequence of random numbers as `scala.util.Random` for a given seed, which is great for testing to make sure we don't have subtle bugs in our random number generator but means there are necessarily going to be significant similarities between our implementations. The underlying random number generator is a linear congruential generator based on Knuth's The Art of Computer Programming and the algorithm for generating random Gaussians is the polar method from Box and Muller also described in Knuth. Is it sufficient to describe that or do we need to go back to the drawing board here?

Reply:
Hmm, tough situation. Here are the options I see:

1. A clean room implementation following `Knuth`, without reference to the JDK source code.
2. Possibly another project like [Apache RNG](http://commons.apache.org/proper/commons-math/javadocs/api-3.6.1/org/apache/commons/math3/random/package-summary.html) has something already pre-written (also [here](https://github.com/encog/encog-java-core/blob/master/src/main/java/org/encog/mathutil/randomize/generate/LinearCongruentialRandom.java), [here](https://www.rosettacode.org/wiki/Knuth_shuffle), [here](https://typelevel.org/cats/datatypes/state.html))

Reply:
What if we wrap `java.util.Random` in a `private` `Ref`? I think I can give us the same interface and hide the mutability. Otherwise I can do the second option.

Reply:
That should work!

Reply:
Oh wait, it will not be purely functional. Let me see if there's a hook we can use to get access into the (deterministic) random generator inside `java.util.Random`.

Reply:
Oh, I just found this:

<https://github.com/scala-js/scala-js/blob/master/javalib/src/main/scala/java/util/Random.scala>

Still looking though...

Reply:
All right, we can [get the seed](https://gist.github.com/wookietreiber/cb9117c9fb747be3fa270242302bd073) on the JVM using reflection.

This means we could construct a `new Random(seed)`, get the next random, get the new seed, and update the seed, all atomically inside a `Ref`.

So the `Ref` approach (delegating to `java.util.Random`) would be viable for JVM.

It would not work for Scala.js, though. 

I think our best option may be basing our work off the Scala.js implementation, which is Apache 2 licensed, and then providing proper attribution and copyright notice on that file. 

What do you think?

Reply:
Yes I think that makes sense. The Scala.js implementation is very similar to our existing one. The main difference is they break the seed down into two 24 bit parts instead of storing the seed as a 48 bit long, but the code for methods like `nextBytes` and `nextGaussian` is the same.

Reply:
Looking from [the history](https://github.com/scala-js/scala-js/commit/4053f3a18fc7efc6c96b117bfc9fff6c7852dec6#diff-29a10fbf90dbeb23b7016d48d5411d99) it looks to be an original implementation; and if it's licensed with Apache 2, then it cannot be derived from the GPL implementation. Of course the math is the same, no way around that, but previous rulings of courts have decided in cases where there is only one "obvious" way to produce a desired result, a novel implementation is not subject to copyright infringement.

Reply:
Thanks for your guidance on this!

Reply:
Happy to help!